File: validate.js
import Ajv2020 from 'ajv/dist/2020.js';
import addFormats from 'ajv-formats';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

try {

  // Use Ajv2020 for draft-2020-12 support
  const ajv = new Ajv2020({
    allErrors: true,
    strict: false,
    verbose: true
  });
  addFormats(ajv);

  // Helper to load a schema
  function loadSchema(relPath) {
    try {
      const fullPath = path.join(__dirname, relPath);
      const content = fs.readFileSync(fullPath, 'utf8');
      const schema = JSON.parse(content);
      console.log(`Loaded schema: ${relPath}`);
      return schema;
    } catch (e) {
      console.error(`Failed to load schema: ${relPath}`);
      console.error(e.message);
      throw e;
    }
  }

  // Helper to resolve schema references
  function resolveSchemaRef(basePath, ref) {
    if (!ref) return null;
    
    console.log(`Resolving reference: ${ref} from ${basePath}`);
    
    // Handle relative paths
    if (ref.startsWith('./') || ref.startsWith('../')) {
      // Always use the __dirname as base and the basePath directory
      const baseDir = path.join(__dirname, path.dirname(basePath));
      const resolved = path.resolve(baseDir, ref);
      console.log(`Resolved relative path to: ${resolved}`);
      return path.relative(__dirname, resolved); // Return relative to project root
    }
    
    // Handle absolute paths (from $id)
    if (ref.startsWith('immerschema/')) {
      const resolved = ref.replace('immerschema/', '');
      console.log(`Resolved immerschema path to: ${resolved}`);
      return resolved;
    }
    
    // Handle direct file references
    if (ref.endsWith('.json')) {
      console.log(`Resolved direct file reference to: ${ref}`);
      return ref;
    }
    
    return null;
  }

  // Helper to register a schema and its dependencies recursively
  const registeredSchemas = new Set();
  
  function registerSchema(schemaPath) {
    // Avoid infinite recursion and duplicate registrations
    if (registeredSchemas.has(schemaPath)) {
      console.log(`Schema already registered: ${schemaPath}`);
      return;
    }
    
    try {
      const schema = loadSchema(schemaPath);
      registeredSchemas.add(schemaPath);
      
      // Register the schema itself
      try {
        ajv.addSchema(schema, schema.$id || schemaPath);
        console.log(`Registered schema: ${schema.$id || schemaPath}`);
      } catch (e) {
        if (e.message.includes('already exists')) {
          console.log(`Schema already registered: ${schema.$id || schemaPath}`);
        } else {
          throw e;
        }
      }
      
      // Handle direct $ref
      if (schema.$ref) {
        const refPath = resolveSchemaRef(schemaPath, schema.$ref);
        if (refPath && fs.existsSync(refPath)) {
          registerSchema(refPath);
        } else if (refPath) {
          console.warn(`Reference not found: ${refPath}`);
          console.warn(`  Referenced from: ${schemaPath}`);
          console.warn(`  Reference: ${schema.$ref}`);
        }
      }
      
      // Handle allOf, anyOf, oneOf arrays
      ['allOf', 'anyOf', 'oneOf'].forEach(key => {
        if (Array.isArray(schema[key])) {
          schema[key].forEach(subSchema => {
            if (subSchema.$ref) {
              const refPath = resolveSchemaRef(schemaPath, subSchema.$ref);
              if (refPath && fs.existsSync(refPath)) {
                registerSchema(refPath);
              } else if (refPath) {
                console.warn(`Reference not found: ${refPath}`);
                console.warn(`  Referenced from: ${schemaPath}`);
                console.warn(`  Reference: ${subSchema.$ref}`);
              }
            }
          });
        }
      });

      // Handle properties that might contain $refs
      if (schema.properties) {
        Object.entries(schema.properties).forEach(([propName, prop]) => {
          if (prop.$ref) {
            const refPath = resolveSchemaRef(schemaPath, prop.$ref);
            if (refPath && fs.existsSync(refPath)) {
              registerSchema(refPath);
            } else if (refPath) {
              console.warn(`Reference not found: ${refPath}`);
              console.warn(`  Referenced from: ${schemaPath}`);
              console.warn(`  Property: ${propName}`);
              console.warn(`  Reference: ${prop.$ref}`);
            }
          }
        });
      }
    } catch (e) {
      console.error(`Failed to register schema: ${schemaPath}`);
      console.error(`Error details: ${e.message}`);
      if (e.stack) {
        console.error(`Stack trace: ${e.stack}`);
      }
    }
  }

  // Register all base schemas first
  console.log('\nRegistering base schemas...');
  const baseSchemas = [
    'node_modules/immerschema/schemas/enum/technique.enum.json',
    'node_modules/immerschema/schemas/enum/riskflag.enum.json',
    'node_modules/immerschema/schemas/enum/screenzone.enum.json',
    'node_modules/immerschema/schemas/enum/software.enum.json',
    'node_modules/immerschema/schemas/enum/role.enum.json',
    'node_modules/immerschema/schemas/ext/debug.schema.json'
  ];
  
  baseSchemas.forEach(schemaPath => {
    registerSchema(schemaPath);
  });

  // Register slice schemas
  console.log('\nRegistering slice schemas...');
  const sliceSchemas = [
    'node_modules/immerschema/schemas/slices/id.slice.schema.json',
    'node_modules/immerschema/schemas/slices/note.slice.schema.json',
    'node_modules/immerschema/schemas/slices/timing-seconds.slice.schema.json',
    'node_modules/immerschema/schemas/slices/timing-frames.slice.schema.json',
    'node_modules/immerschema/schemas/slices/timing.slice.schema.json',
    'node_modules/immerschema/schemas/slices/tech-group.slice.schema.json',
    'node_modules/immerschema/schemas/slices/technique.slice.schema.json',
    'node_modules/immerschema/schemas/slices/software.slice.schema.json',
    'node_modules/immerschema/schemas/slices/tasks.slice.schema.json',
    'node_modules/immerschema/schemas/slices/crew.slice.schema.json',
    'node_modules/immerschema/schemas/slices/risk.slice.json',
    'node_modules/immerschema/schemas/slices/voice.slice.schema.json',
    'node_modules/immerschema/schemas/slices/meta-scene.slice.schema.json',
    'node_modules/immerschema/schemas/slices/screen.slice.schema.json',
    'node_modules/immerschema/schemas/slices/assets.slice.schema.json',
    'node_modules/immerschema/schemas/slices/description.slice.schema.json',
    'node_modules/immerschema/schemas/slices/workload.slice.json',
    'node_modules/immerschema/schemas/slices/bottleneck.slice.json',
    'node_modules/immerschema/schemas/slices/audio.slice.schema.json'
  ];
  
  sliceSchemas.forEach(schemaPath => {
    registerSchema(schemaPath);
  });

  // Register profile schemas FIRST (before core schema that references them)
  const profileSchemas = [
    'node_modules/immerschema/schemas/profiles/draft.schema.json',
    'node_modules/immerschema/schemas/profiles/review.schema.json',
    'node_modules/immerschema/schemas/profiles/plan.schema.json',
    'node_modules/immerschema/schemas/profiles/assign.schema.json',
    'node_modules/immerschema/schemas/profiles/lock.schema.json'
  ];
  
  console.log('\nRegistering profile schemas...');
  profileSchemas.forEach(schemaPath => {
    registerSchema(schemaPath);
  });

  // Load and register project schema (depends on profiles)
  console.log('\nRegistering project schema...');
  const projectSchema = loadSchema('node_modules/immerschema/schemas/project.draft.schema.json');
  ajv.addSchema(projectSchema, projectSchema.$id);

  // Helper to validate a file against a schema
  function validateFile(filePath, schemaPath, description) {
    try {
      console.log(`\nValidating ${description}...`);
      
      // Check if test file exists
      if (!fs.existsSync(filePath)) {
        console.log(`⚠️  Test file not found: ${filePath} - SKIPPING`);
        return null; // Return null to indicate skipped
      }
      
      const data = loadSchema(filePath);
      const schema = loadSchema(schemaPath);
      // Use already-registered schema by $id or path
      const validate = ajv.getSchema(schema.$id || schemaPath);
      if (!validate) {
        console.error(`❌ No validator found for schema: ${schema.$id || schemaPath}`);
        return false;
      }
      const valid = validate(data);
      
      if (valid) {
        console.log(`✅ ${description} validation: PASSED`);
      } else {
        console.log(`❌ ${description} validation: FAILED`);
        console.log('Validation errors:', JSON.stringify(validate.errors, null, 2));
      }
      return valid;
    } catch (e) {
      console.error(`❌ ${description} validation: ERROR`);
      console.error(e.message);
      return false;
    }
  }

  // Validate test files
  const results = {
    passed: 0,
    failed: 0,
    skipped: 0,
    total: 0
  };

  // Validate shots against different profiles
  const shotFiles = [
    { file: 'test_shot.draft.json', profile: 'node_modules/immerschema/schemas/profiles/draft.schema.json', desc: 'Draft shot' },
    { file: 'test_shot.review.json', profile: 'node_modules/immerschema/schemas/profiles/review.schema.json', desc: 'Review shot' },
    { file: 'test_shot.plan.json', profile: 'node_modules/immerschema/schemas/profiles/plan.schema.json', desc: 'Plan shot' },
    { file: 'test_shot.assign.json', profile: 'node_modules/immerschema/schemas/profiles/assign.schema.json', desc: 'Assign shot' },
    { file: 'test_shot.lock.json', profile: 'node_modules/immerschema/schemas/profiles/lock.schema.json', desc: 'Lock shot' }
  ];

  shotFiles.forEach(({ file, profile, desc }) => {
    results.total++;
    const result = validateFile(file, profile, desc);
    if (result === null) {
      results.skipped++;
    } else if (result) {
      results.passed++;
    } else {
      results.failed++;
    }
  });

  // Validate project
  results.total++;
  const projectResult = validateFile('test_project.json', 'node_modules/immerschema/schemas/project.draft.schema.json', 'Project');
  if (projectResult === null) {
    results.skipped++;
  } else if (projectResult) {
    results.passed++;
  } else {
    results.failed++;
  }

  // Print summary
  console.log('\n📊 Final Results');
  console.log('=================');
  console.log(`✅ Passed: ${results.passed}`);
  console.log(`❌ Failed: ${results.failed}`);
  console.log(`⚠️  Skipped: ${results.skipped}`);
  console.log(`📈 Total: ${results.total}`);

} catch (err) {
  console.error('Unexpected error:', err);
  process.exit(1);
} 

File: validate.py
import json
import os
import time
from jsonschema import validate
import copy
import logging

# Set up logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Get the current directory
current_dir = os.path.dirname(os.path.abspath(__file__))
schemas_dir = os.path.join(current_dir, 'schemas')
test_dir = os.path.join(current_dir, 'test')

def load_json_file(filename):
    """Load a JSON file from the test directory"""
    try:
        filepath = os.path.join(test_dir, filename)
        with open(filepath, 'r', encoding='utf-8') as f:
            return json.load(f)
    except Exception as e:
        logger.error(f"Error loading {filename}: {str(e)}")
        return None

def load_schema_file(relative_path):
    """Load a schema file from the schemas directory"""
    try:
        filepath = os.path.join(schemas_dir, relative_path)
        with open(filepath, 'r', encoding='utf-8') as f:
            return json.load(f)
    except Exception as e:
        logger.error(f"Error loading schema {relative_path}: {str(e)}")
        return None

def resolve_schema_refs(schema):
    """Recursively resolve $ref references in a schema by replacing them with actual content"""
    
    # Load all enum schemas into a cache
    enum_cache = {}
    enum_files = ['screenzone.enum.json', 'technique.enum.json', 'software.enum.json', 
                  'role.enum.json', 'riskflag.enum.json']
    for enum_file in enum_files:
        content = load_schema_file(f'enum/{enum_file}')
        if content:
            enum_cache[f"./enum/{enum_file}"] = content
    
    # Load ext schemas
    ext_cache = {}
    ext_files = ['debug.schema.json']
    for ext_file in ext_files:
        content = load_schema_file(f'ext/{ext_file}')
        if content:
            ext_cache[f"./ext/{ext_file}"] = content
    
    # Load core schema
    core_cache = {}
    core_content = load_schema_file('core.schema.json')
    if core_content:
        # Pre-resolve references in the core schema itself
        resolved_core = resolve_refs_in_schema(core_content, enum_cache, ext_cache, {})
        core_cache["./core.schema.json"] = resolved_core
    
    def resolve_refs_recursive(obj):
        if isinstance(obj, dict):
            if '$ref' in obj:
                ref_path = obj['$ref']
                # Handle local file references
                if ref_path in enum_cache:
                    return enum_cache[ref_path]
                elif ref_path in ext_cache:
                    return ext_cache[ref_path]
                elif ref_path in core_cache:
                    return core_cache[ref_path]
                elif ref_path.startswith('./ext/debug.schema.json#/properties/debug'):
                    # Handle specific property reference in debug schema
                    if './ext/debug.schema.json' in ext_cache:
                        return ext_cache['./ext/debug.schema.json']['properties']['debug']
                elif ref_path.startswith('#/'):
                    # Handle internal references (leave as is for now)
                    return obj
                else:
                    logger.warning(f"Unresolved reference: {ref_path}")
                    return obj
            else:
                # Recursively process all values in the dictionary
                resolved = {}
                for key, value in obj.items():
                    resolved[key] = resolve_refs_recursive(value)
                return resolved
        elif isinstance(obj, list):
            # Recursively process all items in the list
            return [resolve_refs_recursive(item) for item in obj]
        else:
            # Return primitive values as is
            return obj
    
    return resolve_refs_recursive(copy.deepcopy(schema))

def resolve_refs_in_schema(schema, enum_cache, ext_cache, core_cache):
    """Helper function to resolve references within a schema"""
    
    def resolve_refs_recursive(obj):
        if isinstance(obj, dict):
            if '$ref' in obj:
                ref_path = obj['$ref']
                # Handle local file references
                if ref_path in enum_cache:
                    return enum_cache[ref_path]
                elif ref_path in ext_cache:
                    return ext_cache[ref_path]
                elif ref_path.startswith('./ext/debug.schema.json#/properties/debug'):
                    # Handle specific property reference in debug schema
                    if './ext/debug.schema.json' in ext_cache:
                        return ext_cache['./ext/debug.schema.json']['properties']['debug']
                elif ref_path.startswith('#/'):
                    # Handle internal references (leave as is for now)
                    return obj
                else:
                    return obj
            else:
                # Recursively process all values in the dictionary
                resolved = {}
                for key, value in obj.items():
                    resolved[key] = resolve_refs_recursive(value)
                return resolved
        elif isinstance(obj, list):
            # Recursively process all items in the list
            return [resolve_refs_recursive(item) for item in obj]
        else:
            # Return primitive values as is
            return obj
    
    return resolve_refs_recursive(copy.deepcopy(schema))

def validate_file(data, schema, description):
    """Validate data against a schema"""
    try:
        validate(instance=data, schema=schema)
        logger.info(f"✅ {description} validation: PASSED")
        return True
    except Exception as e:
        logger.error(f"❌ {description} validation: FAILED")
        logger.error(f"Error: {str(e)}")
        return False

def main():
    # Load test data
    test_shot = load_json_file('test_shot.json')
    test_project = load_json_file('test_project.json')
    validation_tests = load_json_file('validation_tests.json')
    edge_cases = load_json_file('edge_cases.json')

    # Load schemas
    core_schema = load_schema_file('core.schema.json')
    project_schema = load_schema_file('project.schema.json')

    if not all([core_schema, project_schema]):
        logger.error("Failed to load required schema files")
        return

    # Resolve references in schemas
    resolved_core_schema = resolve_schema_refs(core_schema)
    resolved_project_schema = resolve_schema_refs(project_schema)

    # Initialize results
    results = {
        'passed': 0,
        'failed': 0,
        'total': 0
    }

    # Validate test files
    if test_shot:
        results['total'] += 1
        if validate_file(test_shot, resolved_core_schema, "Shot"):
            results['passed'] += 1
        else:
            results['failed'] += 1

    if test_project:
        results['total'] += 1
        if validate_file(test_project, resolved_project_schema, "Project"):
            results['passed'] += 1
        else:
            results['failed'] += 1

    # Print summary
    logger.info("\n📊 Final Results")
    logger.info("=================")
    logger.info(f"✅ Passed: {results['passed']}")
    logger.info(f"❌ Failed: {results['failed']}")
    logger.info(f"📈 Total: {results['total']}")

if __name__ == "__main__":
    main() 